File: src/agent/tools/mcp/oauth.py
"""OAuth browser flow support for MCP HTTP servers."""

from __future__ import annotations

import json
import threading
import webbrowser
from asyncio import Event, get_running_loop
from http.server import BaseHTTPRequestHandler, HTTPServer
from pathlib import Path
from typing import Any
from urllib.parse import parse_qs, urlparse

from mcp.shared.auth import OAuthClientInformationFull, OAuthToken


class FileTokenStorage:
    """Persist OAuth tokens and client info to disk."""

    def __init__(self, server_name: str, base_dir: Path | None = None) -> None:
        self._dir = (base_dir or Path(".agent/oauth")) / server_name
        self._dir.mkdir(parents=True, exist_ok=True)

    def _tokens_path(self) -> Path:
        return self._dir / "tokens.json"

    def _client_path(self) -> Path:
        return self._dir / "client.json"

    async def get_tokens(self) -> OAuthToken | None:
        path = self._tokens_path()
        if not path.exists():
            return None
        try:
            data = json.loads(path.read_text())
            return OAuthToken(**data)
        except (json.JSONDecodeError, Exception):
            return None

    async def set_tokens(self, tokens: OAuthToken) -> None:
        self._tokens_path().write_text(tokens.model_dump_json(indent=2))

    async def get_client_info(self) -> OAuthClientInformationFull | None:
        path = self._client_path()
        if not path.exists():
            return None
        try:
            data = json.loads(path.read_text())
            return OAuthClientInformationFull(**data)
        except (json.JSONDecodeError, Exception):
            return None

    async def set_client_info(self, client_info: OAuthClientInformationFull) -> None:
        self._client_path().write_text(client_info.model_dump_json(indent=2))


CALLBACK_PORT = 18247  # Arbitrary high port for OAuth callback


async def open_browser_redirect(url: str) -> None:
    """Open the authorization URL in the user's default browser."""
    print("Opening browser for authentication...")
    webbrowser.open(url)


async def wait_for_callback() -> tuple[str, str | None]:
    """Run a temporary local HTTP server to receive the OAuth callback.

    Returns (auth_code, state).
    """
    loop = get_running_loop()
    done = Event()
    result: dict[str, Any] = {}

    class Handler(BaseHTTPRequestHandler):
        def do_GET(self) -> None:
            qs = parse_qs(urlparse(self.path).query)
            result["code"] = qs.get("code", [None])[0]
            result["state"] = qs.get("state", [None])[0]

            self.send_response(200)
            self.send_header("Content-Type", "text/html")
            self.end_headers()
            self.wfile.write(
                b"<html><body><h2>Authentication successful!</h2>"
                b"<p>You can close this tab and return to the agent.</p>"
                b"<script>window.close()</script></body></html>"
            )
            loop.call_soon_threadsafe(done.set)

        def log_message(self, format: str, *args: Any) -> None:
            pass  # Suppress HTTP server logs

    server = HTTPServer(("127.0.0.1", CALLBACK_PORT), Handler)

    def _serve_until_done() -> None:
        while not done.is_set():
            server.handle_request()
        server.server_close()

    # Run blocking server in a thread
    thread = threading.Thread(target=_serve_until_done, daemon=True)
    thread.start()

    # Wait for the callback asynchronously
    await done.wait()

    code = result.get("code", "")
    state = result.get("state")
    if not code:
        raise RuntimeError("No authorization code received from OAuth callback")
    return code, state

File: src/tools/mcp/oauth.ts
/** OAuth browser flow support for MCP HTTP servers (file-backed token storage). */

import fs from 'node:fs';
import http from 'node:http';
import path from 'node:path';
import { URL } from 'node:url';

export const CALLBACK_PORT = 18247;

/**
 * An OAuthClientProvider (duck-typed against @modelcontextprotocol/sdk) that
 * persists tokens/client info to disk and drives a browser-based auth flow.
 */
export class FileOAuthProvider {
  private dir: string;

  constructor(
    private serverName: string,
    baseDir = '.agent/oauth',
  ) {
    this.dir = path.join(baseDir, serverName);
    fs.mkdirSync(this.dir, { recursive: true });
  }

  get redirectUrl(): string {
    return `http://127.0.0.1:${CALLBACK_PORT}/callback`;
  }

  get clientMetadata(): Record<string, any> {
    return {
      client_name: `agent-${this.serverName}`,
      redirect_uris: [this.redirectUrl],
      grant_types: ['authorization_code', 'refresh_token'],
      response_types: ['code'],
      token_endpoint_auth_method: 'client_secret_basic',
    };
  }

  private readJson(file: string): any {
    const p = path.join(this.dir, file);
    if (!fs.existsSync(p)) return undefined;
    try {
      return JSON.parse(fs.readFileSync(p, 'utf-8'));
    } catch {
      return undefined;
    }
  }

  private writeJson(file: string, data: unknown): void {
    fs.writeFileSync(path.join(this.dir, file), JSON.stringify(data, null, 2));
  }

  clientInformation(): any {
    return this.readJson('client.json');
  }

  saveClientInformation(info: any): void {
    this.writeJson('client.json', info);
  }

  tokens(): any {
    return this.readJson('tokens.json');
  }

  saveTokens(tokens: any): void {
    this.writeJson('tokens.json', tokens);
  }

  saveCodeVerifier(verifier: string): void {
    fs.writeFileSync(path.join(this.dir, 'verifier.txt'), verifier);
  }

  codeVerifier(): string {
    return fs.readFileSync(path.join(this.dir, 'verifier.txt'), 'utf-8');
  }

  async redirectToAuthorization(authorizationUrl: URL): Promise<void> {
    console.error('Opening browser for authentication...');
    const { default: open } = await import('open');
    await open(authorizationUrl.toString());
  }
}

/**
 * Run a temporary local HTTP server to receive the OAuth callback.
 * Resolves with [authCode, state].
 */
export function waitForCallback(): Promise<[string, string | null]> {
  return new Promise((resolve, reject) => {
    const server = http.createServer((req, res) => {
      const url = new URL(req.url ?? '/', `http://127.0.0.1:${CALLBACK_PORT}`);
      const code = url.searchParams.get('code');
      const state = url.searchParams.get('state');

      res.writeHead(200, { 'Content-Type': 'text/html' });
      res.end(
        '<html><body><h2>Authentication successful!</h2>' +
          '<p>You can close this tab and return to the agent.</p>' +
          '<script>window.close()</script></body></html>',
      );
      server.close();
      if (!code) {
        reject(new Error('No authorization code received from OAuth callback'));
      } else {
        resolve([code, state]);
      }
    });
    server.on('error', reject);
    server.listen(CALLBACK_PORT, '127.0.0.1');
  });
}
